prototype int8range

michae2 opened the pull request:
We're building a prototype of [range types](https://www.postgresql.org/docs/16/rangetypes.html) for breather week! Starting with `int8range`.

Informs: #27791

Commit:
Merge #128400

128400: sqlstats/insights: fix mem leaks on session close r=xinhaoz a=xinhaoz

insights: ensure releasing to Insight pool clears slice

Ensure that when we return objects into the Insight pool
that we release the statmeents in the slice. This fixes
an issue in the logic that returned this object to the pool
which did not nil the `Statements` slice, making it possible
for these slices to grow in capacity in the node's lifetime,
holding onto garbage Statement objects.

This is a lead up to https://github.com/cockroachdb/cockroach/issues/128199 which will likely remove
this pool and reuse the existing statmentBuf pool.

Epic: none
Release note: None

insights: add testing knobs

Add insights specific testing knobs. We'll add some
knobs in later commts.

Epic: none
Release note: None

sqlstats/insights: free memory allocated per session on session close

The insights locking registry buffers statement insight objects
by session id until receiving a transaction insight, when the
buffer is emptied. These buffers can leak if the session is closed
midway through a transaction since the registry will never
receive a transaction event for the session. This commit ensures
we clear any memory allocated in insights for a session by
sending an event to clear the container's session entry, if it
exists, on session close.

A testing knob was added, OnCloseSession, which is called
when the locking registry clears a session.

Epic: none
Fixes: https://github.com/cockroachdb/cockroach/issues/128213

Release note (bug fix): Fixes a memory leak where statement
insight objects may leak if the session is closed without
the transaction finishing.

insights: move insights testing knobs from sqlstats

Move some insights testing knobs that were on the
sqlstats testing knobs to insights pkg.

Epic: none

Release note: None

Co-authored-by: Xin Hao Zhang <xzhang@cockroachlabs.com>

ZhouXing19 committed:
hello from jane

sambhav-jain-16 committed:
workload: empty

Epic: none

Release note: None

michae2 committed:
range types!!

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129233)
<!-- Reviewable:end -->

michae2 committed:
Merge branch 'rangetypes' of github.com:michae2/cockroach into rangetypes

michae2 commented:
@ZhouXing19 and @sambhav-jain-16, it's a bit of a hack, but I added you both as reviewers to this fake PR so that hopefully you'll get email notifications whenever someone pushes to the branch

ZhouXing19 committed:
prelimilary draft for types.T, datum and builtin

ZhouXing19 committed:
support bound fmt in the builtin

michae2 committed:
stub valueside encoding

sambhav-jain-16 committed:
add pg wire

ZhouXing19 committed:
add fmt func for range

ZhouXing19 committed:
allow null

ZhouXing19 committed:
added other methods except compare

michae2 committed:
add ability to cast from string to int8range and vice versa

ZhouXing19 committed:
draft for intersection

ZhouXing19 committed:
encoding, intersect

sambhav-jain-16 committed:
make string parsing working

sambhav-jain-16 committed:
add range merge function

sambhav-jain-16 committed:
fix int8range(a,b) statement

michae2 committed:
whoops, remove int8range from the grammar

sambhav-jain-16 committed:
make parsing consistent

michae2 committed:
Revert "make parsing consistent"

This reverts commit 82f4fb1a443e0456247a94245364698f9eb802b8.

sambhav-jain-16 committed:
make int8range() and ::int8range consistent

michae2 committed:
normalization and cmp

sambhav-jain-16 committed:
add lower and upper functions

sambhav-jain-16 committed:
add isempty function and add lower bound less than upper bound check

sambhav-jain-16 committed:
use datum isEmpty for isempty builtin

sambhav-jain-16 committed:
add lower_inf and upper_inf functions

sambhav-jain-16 committed:
fix create,show columns and insert commands

sambhav-jain-16 committed:
fix select column from table command

sambhav-jain-16 committed:
fix decoding of null values

sambhav-jain-16 committed:
fix error

sambhav-jain-16 committed:
fix row printing in test

ZhouXing19 committed:
ContainedBy and Contains

michae2 committed:
add IN, IS DISTINCT FROM, implicit casting from string constants, cast builtins

sambhav-jain-16 committed:
add overlap operation

sambhav-jain-16 committed:
add adjacent operation

sambhav-jain-16 committed:
fix empty decoding

sambhav-jain-16 committed:
add strict left and strict right operators

sambhav-jain-16 committed:
add union operator

ZhouXing19 committed:
support intersect op

ZhouXing19 committed:
diff, not intensively tested

ZhouXing19 committed:
6 bound typs, func cleanups

ZhouXing19 committed:
updated enc/dec, fwd index SEEMS working

You might see panic after a while after executing:

root@localhost:26259/defaultdb> CREATE TABLE ab (a INT PRIMARY KEY, b INT8RANGE);
CREATE TABLE

Time: 21ms total (execution 21ms / network 1ms)

root@localhost:26259/defaultdb> CREATE INDEX ON ab (b);
CREATE INDEX

Time: 652ms total (execution 652ms / network 0ms)

root@localhost:26259/defaultdb> INSERT INTO ab VALUES (1, '[2,3)'::int8range);
INSERT 0 1

michae2 committed:
overleft and overright ops

michae2 committed:
Revert "updated enc/dec, fwd index SEEMS working"

This reverts commit 74bc5164521c98582f6428efc98a88de82e9a45a.
